src/v/storage/disk_log_impl.cc

```diff
@@ -256,10 +259,104 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
+     *
+     * the strategy is to choose a pair of adjacent segments and first combine
+     * them into a single segment that replaces the pair, and then perform
+     * self-compaction on the replacement segment.
+     */
+    if (_segs.size() < 2) {
+        return std::nullopt;
+    }
+
+    // sliding window over segments. currently restricted to two segments
+    auto range = std::make_pair(_segs.begin(), std::next(_segs.begin(), 2));
+
+    do {
+        // the simple compaction process in use right now builds a concatenation
+        // of segments so we avoid processing a group that is too large.
+        auto size = std::accumulate(
+          range.first,
+          range.second,
+          size_t(0),
+          [](size_t acc, ss::lw_shared_ptr<segment>& seg) {
+              return acc + seg->size_bytes();
+          });
+        if (size < _manager.config().max_compacted_segment_size) {
+            break;
+        }
+
+        // advance the window
+        ++range.first;
+        ++range.second;
+    } while (range.second != _segs.end());
+
+    // no candidate pair found
+    if (range.second == _segs.end()) {
+        return std::nullopt;
+    }
+
+    // the chosen segments all need to be stable
+    const auto unstable = std::any_of(
+      range.first, range.second, [](ss::lw_shared_ptr<segment>& seg) {
+          return seg->has_appender();
+      });
+    if (unstable) {
+        return std::nullopt;
+    }
+
+    return range;
+}
+
+ss::future<> disk_log_impl::compact_adjacent_segments(
+  std::pair<segment_set::iterator, segment_set::iterator> range,
+  storage::compaction_config cfg) {
+    // lightweight copy of segments in range
+    std::vector<ss::lw_shared_ptr<segment>> segments;
+    std::copy(range.first, range.second, std::back_inserter(segments));
+
+    auto target = segments.front();
+
+    /*
+     * remove index files. they will be rebuilt by the single segment compaction
+     * operation, and also ensures we examine segments correctly on recovery.
+     */
+    if (co_await ss::file_exists(target->index().filename())) {
+        co_await ss::remove_file(target->index().filename());
+    }
+
+    auto compact_index = internal::compacted_index_path(
+      target->reader().filename().c_str());
+    if (co_await ss::file_exists(compact_index.string())) {
+        co_await ss::remove_file(compact_index.string());
+    }
+
+    // combine segments in the given segment set range
+    co_await storage::internal::collapse_segments(segments, cfg, _probe);
+
+    // compact the combined segment
+    target->clear_finished_self_compaction();
+    co_await storage::internal::self_compact_segment(target, cfg, _probe);
+    target->mark_as_finished_self_compaction();
+
+    // remove redundant range
+    _segs.erase(std::next(range.first), range.second);
+    for (auto it = std::next(segments.begin()); it != segments.end(); it++) {
+        co_await remove_segment_permanently(*it, "compact_adjacent_segments");
```

Comment:
interesting. I think should have a comment that while now there is only a for-loop of size 1, this will for for any range.